ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala

```diff
@@ -0,0 +1,183 @@
+package ch.epfl.ts.data
+
+import ch.epfl.ts.data.Currency.Currency
+import scala.reflect.ClassTag
+
+
+/**
+ * Class behaving much like a map, designed to hold values
+ * for a trading strategy's parameters.
+ * Each parameter checks the validity of its value on instantiation.
+ * 
+ * It can be constructed with an arbitrary number of parameters.
+ */
+class StrategyParameters(params: (String, Parameter[_])*) {
+  type Key = String
+  val parameters = params.toMap
+  
+  /**
+   * Similar to what a map's `get` would do.
+   * We also perform type checking.
+   */
+  def get[T : ClassTag](key: Key): Option[T] = {
+    parameters.get(key) match {
+      case Some(p) => p.get() match {
+        case v: T => Some(v)
+        case _ => None
+      }
+      case _ => None
+    }
+  }
+  
+  /**
+   * Get the value for this key
+   * or fallback on the provided default if:
+   *   - Such a key doesn't exist
+   *   - The key exists but doesn't have the right parameter type
+   */
+  def getOrElse[T : ClassTag](key: Key, fallback: T): T =
+    get[T](key) match {
+      case Some(v) => v
+      case _ => fallback
+    }
+  
+  def getOrDefault[T : ClassTag](key: Key, parameterType: ParameterTrait[T]): T =
+    getOrElse(key, parameterType.defaultValue)
+  
+  override def toString: String = {
+    val strings = (for {
+      p <- parameters
+      key = p._1
+      paramType = p._2.toString()
+      value = p._2.get().toString()
+    } yield key + " (type " + paramType + ") = " + value).toList
```

Comment:
And also, doesn't `p._2.toString()` add the type and value too? Add a small test for that.